tensorboard/plugins/pr_curve/pr_curves_plugin.py

```diff
@@ -0,0 +1,198 @@
+
+    Arguments:
+      runs: A list of runs to fetch the curves for.
+      tag: The tag to fetch the curves for.
+
+    Raises:
+      ValueError: If no PR curves could be fetched for a run and tag.
+
+    Returns:
+      The JSON object for the PR curves route response.
+    """
+    response_mapping = {}
+    for run in runs:
+      try:
+        tensor_events = self._multiplexer.Tensors(run, tag)
+      except KeyError:
+        raise ValueError(
+            'No PR curves could be fetched for run %r and tag %r' % (run, tag))
+
+      response_mapping[run] = [
+          self._process_tensor_event(e) for e in tensor_events]
+    return response_mapping
+
+  @wrappers.Request.application
+  def tags_route(self, request):
+    """A route (HTTP handler) that returns a response with tags.
+    Returns:
+      A response that contains a JSON object. The keys of the object
+      are all the runs. Each run is mapped to a (potentially empty)
+      list of all tags that are relevant to this plugin.
+    """
+    return http_util.Respond(
+        request, self.tags_impl(), 'application/json')
+
+  def tags_impl(self):
+    """Creates the JSON object for the tags route response.
+
+    Returns:
+      The JSON object for the tags route response.
+    """
+    all_runs = self._multiplexer.PluginRunToTagToContent(
+        PrCurvesPlugin.plugin_name)
+    return {
+        run: list(tag_to_content.keys())
+        for (run, tag_to_content) in all_runs.items()
+    }
+
+  @wrappers.Request.application
+  def available_steps_route(self, request):
+    """Gets a dict mapping run to a list of numeric steps.
+
+    Returns:
+      A dict with string keys (all runs with PR curve data). The values of the
+      dict are lists of step values (ints) that should be used for the slider
+      for that run.
+    """
+    return http_util.Respond(
+        request, self.available_steps_impl(), 'application/json')
+
+  def available_steps_impl(self):
+    """Creates the JSON object for the available steps route response.
+
+    Returns:
+      The JSON object for the available steps route response.
+    """
+    all_runs = self._multiplexer.PluginRunToTagToContent(
+        PrCurvesPlugin.plugin_name)
+    response = {}
+
+    # Compute the max step exhibited by any tag for each run. If a run lacks
+    # data, exclude it from the mapping.
+    for run, tag_to_content in all_runs.items():
+      if not tag_to_content:
+        # This run lacks data for this plugin.
+        continue
+
+      # Just use the list of tensor events for any of the tags to determine the
+      # steps to list for the run. The steps are often the same across tags for
+      # each run, albeit the user may elect to sample certain tags differently
+      # within the same run. If the latter occurs, TensorBoard will show the
+      # actual step of each tag atop the card for the tag.
+      tensor_events = self._multiplexer.Tensors(
+          run, list(tag_to_content.keys())[0])
```

Comment:
This is nondeterministic, which rubs me the wrong way. Consider: `min(six.iterkeys(tag_to_content))`.

(General note: If you were to not apply this comment, prefer `next(six.iterkeys(tag_to_content))` to what's written here. No need to serialize to list, duplicate the list (in Python 2), and then grab the first element.)